File: couple-carrier-risk/python/tests/test_config.py
import pytest

from couple_carrier_risk.config import (
    AUTO_ANCESTRY,
    DEFAULT_BASE_URL,
    ConfigError,
    load_config,
)


def _env() -> dict[str, str]:
    return {"EVAGENE_API_KEY": "evg_test"}


def _minimal_argv(*extra: str) -> list[str]:
    return [
        "--partner-a", "a.txt",
        "--partner-b", "b.txt",
        *extra,
    ]


def test_defaults_base_url_format_cleanup_and_auto_ancestry() -> None:
    config = load_config(_minimal_argv(), _env())

    assert config.base_url == DEFAULT_BASE_URL
    assert DEFAULT_BASE_URL == "https://evagene.net"
    assert config.partner_a_file == "a.txt"
    assert config.partner_b_file == "b.txt"
    assert config.ancestry_a == AUTO_ANCESTRY
    assert config.ancestry_b == AUTO_ANCESTRY
    assert config.output_format == "table"
    assert config.cleanup is True


def test_honours_custom_base_url() -> None:
    config = load_config(
        _minimal_argv(),
        {"EVAGENE_API_KEY": "evg_test", "EVAGENE_BASE_URL": "https://evagene.example"},
    )

    assert config.base_url == "https://evagene.example"


@pytest.mark.parametrize("fmt", ["csv", "json"])
def test_accepts_supported_output_formats(fmt: str) -> None:
    config = load_config(_minimal_argv("--output", fmt), _env())

    assert config.output_format == fmt


def test_rejects_unknown_output_format() -> None:
    with pytest.raises(ConfigError):
        load_config(_minimal_argv("--output", "yaml"), _env())


def test_no_cleanup_overrides_default() -> None:
    config = load_config(_minimal_argv("--no-cleanup"), _env())

    assert config.cleanup is False


def test_carries_ancestry_flags() -> None:
    config = load_config(
        _minimal_argv("--ancestry-a", "ashkenazi_jewish", "--ancestry-b", "mediterranean"),
        _env(),
    )

    assert config.ancestry_a == "ashkenazi_jewish"
    assert config.ancestry_b == "mediterranean"


def test_missing_api_key_raises() -> None:
    with pytest.raises(ConfigError, match="EVAGENE_API_KEY"):
        load_config(_minimal_argv(), {})


def test_partner_files_are_required() -> None:
    with pytest.raises(ConfigError):
        load_config(["--partner-a", "a.txt"], _env())

File: couple-carrier-risk/r/tests/testthat/test-config.R
minimal_argv <- function(...) {
  c("--partner-a", "a.txt", "--partner-b", "b.txt", ...)
}

env_with_key <- function(extra = list()) {
  c(list(EVAGENE_API_KEY = "evg_test"), extra)
}

test_that("defaults base URL, output table, auto ancestry, and cleanup on", {
  config <- load_config(minimal_argv(), env_with_key())

  expect_equal(config$base_url, "https://evagene.net")
  expect_equal(config$api_key, "evg_test")
  expect_equal(config$partner_a_file, "a.txt")
  expect_equal(config$partner_b_file, "b.txt")
  expect_equal(config$ancestry_a, "auto")
  expect_equal(config$ancestry_b, "auto")
  expect_equal(config$output, "table")
  expect_true(config$cleanup)
})

test_that("honours a custom base URL", {
  config <- load_config(
    minimal_argv(),
    env_with_key(list(EVAGENE_BASE_URL = "https://evagene.example"))
  )
  expect_equal(config$base_url, "https://evagene.example")
})

test_that("accepts --output csv and --output json", {
  for (fmt in c("csv", "json")) {
    config <- load_config(minimal_argv("--output", fmt), env_with_key())
    expect_equal(config$output, fmt)
  }
})

test_that("rejects unknown --output", {
  expect_error(
    load_config(minimal_argv("--output", "yaml"), env_with_key()),
    class = "config_error"
  )
})

test_that("--no-cleanup overrides the default", {
  config <- load_config(minimal_argv("--no-cleanup"), env_with_key())
  expect_false(config$cleanup)
})

test_that("carries --ancestry-a and --ancestry-b flags", {
  config <- load_config(
    minimal_argv("--ancestry-a", "ashkenazi_jewish", "--ancestry-b", "mediterranean"),
    env_with_key()
  )
  expect_equal(config$ancestry_a, "ashkenazi_jewish")
  expect_equal(config$ancestry_b, "mediterranean")
})

test_that("missing API key raises config_error", {
  expect_error(
    load_config(minimal_argv(), list()),
    class = "config_error"
  )
})

test_that("partner files are required", {
  expect_error(
    load_config(c("--partner-a", "a.txt"), env_with_key()),
    class = "config_error"
  )
})
